decred/tests/unit/dcr/wire/test_msgversion.py

```diff
@@ -0,0 +1,324 @@
+
+    me = netaddress.newNetAddress(tcpAddrMe, wire.SFNodeNetwork)
+    tcpAddrYou = netaddress.TCPAddr(ip="192.168.0.1", port=8333,)
+    you = netaddress.newNetAddress(tcpAddrYou, wire.SFNodeNetwork)
+
+    nonce = random.randint(0, 0xFFFFFFFF)
+
+    # Ensure we get the correct data back out.
+    msg = msgversion.newMsgVersion(me, you, nonce, lastBlock)
+    assert msg.protocolVersion == pver
+
+    assert sameNetAddress(msg.addrMe, me)
+    assert sameNetAddress(msg.addrYou, you)
+    assert msg.nonce == nonce
+    assert msg.userAgent == msgversion.DefaultUserAgent
+    assert msg.lastBlock == lastBlock
+    assert not msg.disableRelayTx
+
+    msg.addUserAgent("myclient", "1.2.3", "optional", "comments")
+    customUserAgent = (
+        msgversion.DefaultUserAgent + "myclient:1.2.3(optional; comments)/"
+    )
+    assert msg.userAgent == customUserAgent
+
+    msg.addUserAgent("mygui", "3.4.5")
+    customUserAgent += "mygui:3.4.5/"
+    assert msg.userAgent == customUserAgent
+
+    # accounting for ":", "/"
+    with pytest.raises(DecredError):
+        msg.addUserAgent(
+            "t" * (msgversion.MaxUserAgentLen - len(customUserAgent) - 2 + 1), ""
+        )
+
+    # Version message should not have any services set by default.
+    assert msg.services == 0
+    assert not msg.hasService(wire.SFNodeNetwork)
+
+    # Ensure the command is expected value.
+    assert msg.command() == "version"
+
+    # Ensure max payload is expected value.
+    # Protocol version 4 bytes + services 8 bytes + timestamp 8 bytes +
+    # remote and local net addresses + nonce 8 bytes + length of user agent
+    # (varInt) + max allowed user agent length + last block 4 bytes +
+    # relay transactions flag 1 byte.
+    wantPayload = 358
+    maxPayload = msg.maxPayloadLength(pver)
+    assert maxPayload == wantPayload
+
+    # Ensure max payload length is not more than MaxMessagePayload.
+    assert maxPayload <= wire.MaxMessagePayload
+
+    # Ensure adding the full service node flag works.
+    msg.addService(wire.SFNodeNetwork)
+    assert msg.services == wire.SFNodeNetwork
+    assert msg.hasService(wire.SFNodeNetwork)
+
+    # Use a fake connection.
+    class fakeConn:
+        localAddr = tcpAddrMe
+        remoteAddr = tcpAddrYou
+
+    msg = msgversion.newMsgVersionFromConn(fakeConn, nonce, lastBlock)
+
+    # Ensure we get the correct connection data back out.
+    assert msg.addrMe.ip == tcpAddrMe.ip
+    assert msg.addrYou.ip == tcpAddrYou.ip
+
+
+def test_VersionWire():
+    """
+    Test the MsgVersion wire encode and decode for various protocol versions.
+    """
+    # verRelayTxFalse and verRelayTxFalseEncoded is a version message as of
+    # BIP0037Version with the transaction relay disabled.
+    baseVersionBIP0037Copy = baseVersionBIP0037
+    verRelayTxFalse = baseVersionBIP0037Copy
+    verRelayTxFalse.disableRelayTx = True
+    verRelayTxFalseEncoded = baseVersionBIP0037Encoded()
+    verRelayTxFalseEncoded[-1] = 0
```

Comment:
none of these are used

Reply:
https://github.com/decred/dcrd/pull/2184